packages/signature-controller/src/SignatureController.ts

```diff
@@ -303,47 +321,121 @@ export class SignatureController extends BaseControllerV2<
    *
    * We currently define our eth_sign and personal_sign mostly for legacy Dapps.
    *
-   * @param msgParams - The params of the message to sign & return to the Dapp.
+   * @param messageParams - The params of the message to sign & return to the Dapp.
    * @param req - The original request, containing the origin.
    * @returns Promise resolving to the raw data of the signature request.
    */
   async newUnsignedPersonalMessage(
-    msgParams: PersonalMessageParams,
+    messageParams: PersonalMessageParams,
     req: OriginalRequest,
   ): Promise<string> {
-    return this.#personalMessageManager.addUnapprovedMessageAsync(
-      msgParams,
+    return this.#newUnsignedAbstractMessage(
+      this.#personalMessageManager,
+      ApprovalType.PersonalSign,
+      'Personal Message',
+      this.#signPersonalMessage.bind(this),
+      messageParams,
       req,
     );
   }
 
   /**
    * Called when a dapp uses the eth_signTypedData method, per EIP 712.
    *
-   * @param msgParams - The params passed to eth_signTypedData.
+   * @param messageParams - The params passed to eth_signTypedData.
    * @param req - The original request, containing the origin.
    * @param version - The version indicating the format of the typed data.
+   * @param signingOpts - An options bag for signing.
+   * @param signingOpts.parseJsonData - Whether to parse the JSON before signing.
    * @returns Promise resolving to the raw data of the signature request.
    */
   async newUnsignedTypedMessage(
-    msgParams: TypedMessageParams,
+    messageParams: TypedMessageParams,
     req: OriginalRequest,
     version: string,
+    signingOpts: TypedMessageSigningOptions,
   ): Promise<string> {
-    return this.#typedMessageManager.addUnapprovedMessageAsync(
-      msgParams,
+    return this.#newUnsignedAbstractMessage(
+      this.#typedMessageManager,
+      ApprovalType.EthSignTypedData,
+      'Typed Message',
+      this.#signTypedMessage.bind(this),
+      messageParams,
+      req,
+      undefined,
       version,
+      signingOpts,
+    );
+  }
+
+  setTypedMessageInProgress(messageId: string) {
+    this.#typedMessageManager.setMessageStatusInProgress(messageId);
+  }
+
+  setPersonalMessageInProgress(messageId: string) {
+    this.#personalMessageManager.setMessageStatusInProgress(messageId);
+  }
+
+  async #newUnsignedAbstractMessage<
+    M extends AbstractMessage,
+    P extends AbstractMessageParams,
+    PM extends AbstractMessageParamsMetamask,
+  >(
+    messageManager: AbstractMessageManager<M, P, PM>,
+    approvalType: ApprovalType,
+    messageName: string,
+    signMessage: (
+      messageParams: PM,
+      version?: string,
+      signingOpts?: TypedMessageSigningOptions,
+    ) => void,
+    messageParams: PM,
+    req: OriginalRequest,
+    validateMessage?: (params: PM) => void,
+    version?: string,
+    signingOpts?: TypedMessageSigningOptions,
```

Comment:
Rather than coupling this type which is specific to typed messages, could we make this a simple template on the method such as `SO`?

Reply:
Done